File: database_connection/task_list/task_list_node/view/HomeView.js
const TaskDto = require("../dto/TaskDto");
const TaskService = require("../service/TaskService");
const question = require("../utils/ReaderUtil");

class HomeView {
  /**
   * @param {TaskService} taskService
   */
  constructor(taskService) {
    this.taskService = taskService;
  }

  async start() {
    const option = await question(`Selecione uma opçao:
1. Criar nova tarefa
2. Visualizar tarefas
3. Atualizar tarefa
4. Excluir tarefa
0. Sair
`);
    await this.handleOption(Number(option));
  }

  async handleOption(option) {
    switch (option) {
      case 1:
        await this.onCreateNewTask();
        break;
      case 2:
        await this.onShowTasks();
        break;
      case 3:
        await this.onUpdateTask();
        break;
      case 4:
        await this.onDeleteTask();
        break;
      case 0:
        this.onExit();
        break;
      default:
        console.log("Opçao inválida. Selecione outra.");
        break;
    }
    await this.start();
  }

  async onCreateNewTask() {
    const title = await question("\nInforme o título da tarefa: ");
    const description = await question("Informe a descriçao da tarefa: ");

    await this.taskService.addNewTask(new TaskDto(title, description));

    console.log("\nTarefa cadastrada com sucesso!");
  }

  async onShowTasks() {
    const tasks = await this.taskService.retrieveAllTasks();

    if (tasks.length) {
      tasks.sort((a, b) => b.created_at - a.created_at);
      console.log("\nTarefas cadastradas: ");
      tasks.forEach((task) => console.log(`Tarefa ${task.id} -> ${task.toString()}`));
      console.log();
    } else {
      console.log("\nNao há nenhuma tarefa cadastrada.\n");
    }
  }

  async onUpdateTask() {
    const selectedTask = await this.selectATask("alterar");

    console.log("\nInforme os dados para alteraçao, caso nao queira mudar, apenas deixe em branco: ");

    const title = await question("\nInforme o título da tarefa: ");
    const description = await question("Informe a descriçao da tarefa: ");

    await this.taskService.updateTask(selectedTask, new TaskDto(title, description));

    console.log("Tarefa alterada com sucesso!");
  }

  async onDeleteTask() {
    const selectedTask = await this.selectATask("deletar");

    const answer = await question(`\nTem certeza que deseja excluir a tarefa ${selectedTask}? (S/N)`);
    if (answer.toLowerCase() == "s") {
      await this.taskService.deleteTaskById(selectedTask);
      console.log("Tarefa deletada com sucesso!");
    } else if (answer.toLowerCase() == "n") {
      console.log("Tarefa não deletada");
    } else {
      console.log("Opçao inválida.");
    }
  }

  async selectATask(operation) {
    const selectedTaskIndex = Number(await question(`\nQual tarefa deseja ${operation}? (Escolha de acordo com o número de cada tarefa)`));
    await this.taskService.retrieveTaskById(selectedTaskIndex); // throws exception if index is out of bounds
    return selectedTaskIndex;
  }

  onExit() {
    process.exit(1);
  }
}

module.exports = HomeView;

File: database_connection/task_list/task_list_java/src/main/java/com/view/HomeView.java
package com.view;

import java.util.List;

import com.dto.TaskDto;
import com.model.Task;
import com.service.TaskService;
import com.utils.MyScanner;

public class HomeView {

  private final MyScanner scanner;
  private final TaskService taskService;

  public HomeView(MyScanner scanner, TaskService taskService) {
    this.scanner = scanner;
    this.taskService = taskService;
  }

  public void start() {
    System.out.println("""
        \nSelecione uma opçao:
        1. Criar nova tarefa
        2. Visualizar tarefas
        3. Atualizar tarefa
        4. Excluir tarefa
        0. Sair """);
    var optionalOption = scanner.readInt();
    handleOption(optionalOption.orElse(5));
  }

  public void handleOption(int option) {
    switch (option) {
      case 1 -> onCreateNewTask();
      case 2 -> onShowTasks();
      case 3 -> onUpdateTask();
      case 4 -> onDeleteTask();
      case 0 -> onExit();
      default -> System.out.println("Opçao inválida. Selecione outra.");
    }
    start();
  }

  private void onCreateNewTask() {
    System.out.print("\nInforme o título da tarefa: ");
    String title = scanner.readLine();

    System.out.print("Informe a descriçao da tarefa: ");
    String description = scanner.readLine();

    taskService.addNewTask(new TaskDto(title, description));

    System.out.println("\nTarefa cadastrada com sucesso!");
  }

  private void onShowTasks() {
    List<Task> tasks = taskService.retrieveAllTasks();

    if (!tasks.isEmpty()) {
      tasks.sort((a, b) -> b.getCreatedAt().compareTo(a.getCreatedAt()));
      System.out.println("\nTarefas cadastradas: ");
      tasks.forEach(task -> System.out.format("Tarefa %d -> %s \n", task.getId(), task.toString()));
    } else {
      System.out.println("\nNao há nenhuma tarefa cadastrada.");
    }
  }

  private void onUpdateTask() {
    int selectedTask = selectATask("alterar") - 1;

    System.out.println("\nInforme os dados para alteraçao, caso nao queira mudar, apenas deixe em branco: ");

    System.out.print("\nInforme o título da tarefa: ");
    String title = scanner.readLine();

    System.out.print("Informe a descriçao da tarefa: ");
    String description = scanner.readLine();

    taskService.updateTask(selectedTask, new TaskDto(title, description));

    System.out.println("Tarefa alterada com sucesso!");
  }

  private void onDeleteTask() {
    int selectedTask = selectATask("deletar");

    System.out.println("\nTem certeza que deseja excluir a tarefa " + selectedTask + "? (S/N)");
    String answer = scanner.readLine();

    if (answer.toLowerCase().equals("s")) {
      taskService.deleteTaskById(selectedTask);
      System.out.println("Tarefa deletada com sucesso!");
    } else if (answer.toLowerCase().equals("n")) {
      System.out.println("Tarefa não deletada");
    } else {
      System.out.println("Opçao inválida.");
    }
  }

  private int selectATask(String operation) {
    System.out.println("\nQual tarefa deseja " + operation + "? (Escolha de acordo com o número de cada tarefa)");
    var selectedTaskIndex = scanner.readInt().orElse(-1);
    taskService.retrieveTaskById(selectedTaskIndex); // throws exception if index is out of bounds
    return selectedTaskIndex;
  }

  private void onExit() {
    System.out.println("Obrigado pela preferência.");
    System.exit(0);
  }
}
